include/irr/asset/IVirtualTexture.h

```diff
@@ -477,141 +638,215 @@ class IVirtualTexture : public core::IReferenceCounted
             bool ok = SMiptailPacker::computeMiptailOffsets(m_miptailOffsets, pgSzLog2, m_tilePadding);
             assert(ok);
         }
+
+        m_addrsArray = core::make_refctd_dynamic_array<decltype(m_addrsArray)>(1u<<(2u*_pgTabSzxy_log2 + 1u));
+        std::fill(m_addrsArray->begin(), m_addrsArray->end(), IVTResidentStorage::phys_pg_addr_alctr_t::invalid_address);
+        m_sizesArray = core::make_refctd_dynamic_array<decltype(m_sizesArray)>(1u<<(2u*_pgTabSzxy_log2 + 1u));
+        std::fill(m_sizesArray->begin(), m_sizesArray->end(), 1u);
     }
 
-    virtual STextureData pack(const image_t* _img, const IImage::SSubresourceRange& _subres, ISampler::E_TEXTURE_CLAMP _wrapu, ISampler::E_TEXTURE_CLAMP _wrapv, ISampler::E_TEXTURE_BORDER_COLOR _borderColor) = 0;
+    SMasterTextureData alloc(E_FORMAT _primaryFormat, const VkExtent3D& _mip0extent, const IImage::SSubresourceRange& _subres, ISampler::E_TEXTURE_CLAMP _wrapu, ISampler::E_TEXTURE_CLAMP _wrapv)
+    {
+        const E_FORMAT format = _primaryFormat;
+        uint32_t smplrIndex = 0u;
+        IVTResidentStorage* storage = nullptr;
+        {
+            auto found = m_storage.find(getFormatClass(format));
+            if (found==m_storage.end())
+                return STextureData::invalid();
+            storage = found->second.get();
+
+            SamplerArray* views = nullptr;
+            if (isFloatingPointFormat(format)||isNormalizedFormat(format)||isScaledFormat(format))
+                views = &m_fsamplers;
+            else if (isSignedFormat(format))
+                views = &m_isamplers;
+            else
+                views = &m_usamplers;
+            auto view_it = std::find_if(views->views->begin(), views->views->end(), [format](const core::smart_refctd_ptr<ICPUImageView>& _view) {return _view->getCreationParameters().format==format;});
+            if (view_it==views->views->end()) //no physical page texture view/sampler for requested format
+                return STextureData::invalid();
+            smplrIndex = std::distance(views->views->begin(), view_it);
+        }
+        auto assignedLayers = getPageTableLayersForFormat(format);
 
-    virtual core::smart_refctd_ptr<image_view_t> createPageTableView() const = 0;
+        const VkExtent3D extent = {_mip0extent.width>>_subres.baseMipLevel, _mip0extent.height>>_subres.baseMipLevel, 1u};
+        uint32_t szAndAlignment = computeSquareSz(extent.width, extent.height);
+        szAndAlignment *= szAndAlignment;
+
+        uint32_t pgtLayer = 0u;
+        uint32_t addr = pg_tab_addr_alctr_t::invalid_address;
+        for (auto it = assignedLayers.first; it != assignedLayers.second; ++it)
+        {
+            pgtLayer = it->second;
+            core::address_allocator_traits<pg_tab_addr_alctr_t>::multi_alloc_addr((*m_pageTableLayerAllocators)[pgtLayer], 1u, &addr, &szAndAlignment, &szAndAlignment, nullptr);
+            if (addr==pg_tab_addr_alctr_t::invalid_address)
+                continue;
```

Comment:
I don't get this `if` + `continue`

Reply:
trying next of layers assigned for this format if previously tried has too little free space to alloc

Reply:
this still doesn't make sense.

If you fail to allocate (get invalid_address), then you go into the `if` and `continue` to do the multi allocation again on the next layer.

if you succeed then you don't hit the `continue` but you go around the loop again anyway!
The only thing that saves you from multiple allocations is that `multi_alloc_addr` doesn't allocate address outputs that were not initialized to `invalid_address`

Despite that, the behaviour still should be 
```
if (addr!=pg_tab_addr_alctr_t::invalid_address)
   break;
````

Reply:
lol, im stupid, yeah, thanks XD ...